### PR TITLE
[Use Cases] add additional use cases for Key in libelektra-core

### PR DESCRIPTION
**Pull request by lawli3t**
## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

File: doc/usecases/core/UC_key_base_name.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+- **Alternative scenario:**
+  - Caller requests to append new name part to `Key`
+  - Core stores appended name in `Key`
+- **Error scenario:**
+  - `Key` has name marked as read-only (e.g., because `Key` is part of ordered collection like `KeySet`)
+  - Caller requests to change basename of `Key` to new basename
+  - Core returns error and basename of `Key` remains unchanged
+- **Error scenario:**
+  - `Key` has name marked as read-only (e.g., because `Key` is part of ordered collection like `KeySet`)
+  - Caller requests to append to name of `Key`
```
Comment (kodebach) [inline doc/usecases/core/UC_key_base_name.md:30]:
```suggestion
  - Caller requests to append new name part to `Key`
```

File: doc/usecases/core/UC_key_clear.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+- **Actors:** Core, Caller
+- **Brief:** Clearing contents of a `Key`
+- **Status:** Implemented
+
+## Scenarios
+
+- **Precondition:**
+  - [`Key` has been created](UC_key_create.md)
+- **Main success scenario:**
+  - Caller requests the fields of a `Key` should get cleard
```
Comment (kodebach) [inline doc/usecases/core/UC_key_clear.md:16]:
```suggestion
  - Caller requests the data of a `Key` should get cleared
```

Data as in "what's publicly visible", some internal fields could remain or be set to specific values (e.g. sync flag).

File: doc/usecases/core/UC_key_copy.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+## Scenarios
+
+- **Precondition:**
+  - [`Key` K1 has been created](UC_key_create.md)
+  - [`Key` K2 has been created](UC_key_create.md)
+- **Main success scenario:**
+  - Caller requests to copy one or multiple fields from `Key` k1 to `Key` k2
+  - Fields that can be copied are name, value and metadata
+  - Fields, that caller wanted to copy, of `Key` k2 now have the value of the respective fields of `Key` k1
```
Comment (kodebach) [inline doc/usecases/core/UC_key_copy.md:19]:
```suggestion
  - Caller requests to copy some or all data (name, value, metadata) from `Key` k1 to `Key` k2
  - The data, that caller wanted to copy, of `Key` k2 now have the value of the respective fields of `Key` k1
```

File: doc/usecases/core/UC_key_lock.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+- **Brief:** Access to and manipulation of `Key`'s lock flags
+- **Status:** Implemented
+
+## Scenarios
+
+- **Precondition:**
+  - [`Key` has been created](UC_key_create.md)
+- **Main success scenario:**
+  - Caller requests lock status of one or multiple flags of a `Key`
+  - Possible lock flags are: name, value, metadata
```
Comment (kodebach) [inline doc/usecases/core/UC_key_lock.md:17]:
```suggestion
  - Caller requests lock status of one or multiple flags (name, value, metadata) of a `Key`
```

File: doc/usecases/core/UC_key_ref.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  - Core returns current value of reference counter of a `Key` after decrementing
+- **Error scenario:**
+  - Caller requests to increment value of reference counter of a `Key` whose reference counter is INT_MAX
+  - Reference counter will not get incremented and stays at INT_MAX
+- **Error scenario:**
+  - Caller requests to decrement value of reference counter of a `Key` whose reference counter is 0
+  - Reference counter will not get decremented and stays at 0
+- **Postcondition:**
+  - Reference Counter has a value from 0 to INT_MAX
+- **Non-functional Constraints:**
```
Comment (kodebach) [inline doc/usecases/core/UC_key_ref.md:32]: The range `0 - INT_MAX` is wrong, all refs are `uint16_t` so it should be

```suggestion
- **Error scenario:**
  - Caller requests to increment value of reference counter of a `Key` whose reference counter is UINT16_MAX
  - Reference counter will not get incremented and stays at UINT16_MAX
- **Error scenario:**
  - Caller requests to decrement value of reference counter of a `Key` whose reference counter is 0
  - Reference counter will not get decremented and stays at 0
- **Postcondition:**
  - Reference Counter has a value from 0 to UINT16_MAX
- **Non-functional Constraints:**
```

**Review by kodebach — CHANGES_REQUESTED** [review]
LGTM, just some minor things

**Review by lukashartl — APPROVED** [review]

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**